arch/src/aarch64/fdt.rs

```diff
@@ -328,8 +375,41 @@ fn create_cpu_nodes(
             fdt.property_u32("i-cache-line-size", l1_i_cache_line_size)?;
             fdt.property_u32("i-cache-sets", l1_i_cache_sets)?;
 
-            if l2_cache_size != 0 {
-                fdt.property_u32("next-level-cache", L2_CACHE_PHANDLE)?;
+            if l2_cache_size != 0 && !l2_cache_shared {
+                fdt.property_u32(
+                    "next-level-cache",
+                    cpu_id as u32 + max_cpus + FIRST_VCPU_PHANDLE + L2_CACHE_PHANDLE,
+                )?;
+
+                let l2_cache_name = "l2-cache0";
+                let l2_cache_node = fdt.begin_node(l2_cache_name)?;
+                fdt.property_u32(
+                    "phandle",
+                    cpu_id as u32 + max_cpus + FIRST_VCPU_PHANDLE + L2_CACHE_PHANDLE,
+                )?;
+
+                fdt.property_string("compatible", "cache")?;
+                fdt.property_u32("cache-size", l2_cache_size)?;
+                fdt.property_u32("cache-line-size", l2_cache_line_size)?;
+                fdt.property_u32("cache-sets", l2_cache_sets)?;
+                fdt.property_u32("cache-level", 2)?;
+
+                if l3_cache_size != 0 && l3_cache_shared {
+                    let package_id: u32 = cpu_id as u32 / cores_per_package as u32;
+                    fdt.property_u32(
+                        "next-level-cache",
+                        package_id
+                            + num_cpus as u32
+                            + max_cpus
+                            + FIRST_VCPU_PHANDLE
+                            + L2_CACHE_PHANDLE
+                            + L3_CACHE_PHANDLE,
+                    )?;
+                }
+
+                fdt.end_node(l2_cache_node)?;
+            } else if l2_cache_size != 0 && l2_cache_shared {
+                error!("L2 cache shared with other cpus");
```

Comment:
Our SOC(with ARMv8 cores) has L2 cache shared among a set of cores. Does this change break on our system when I launch a VM?.

`index2# pwd
/sys/devices/system/cpu/cpu0/cache/index2
index2# cat shared_cpu_list
0,4,8,12
index2# cat level
2`

Reply:
Thanks for the comment, it's a good question. In fact, if the affinity between the VCPU and the Host CPU is not set, it is difficult for us to determine the L2 cache cache sharing relationship of different CPUs in the Guest VCPU, and the VCPU may migrate between different CPUs. Therefore, we assume that the L2 cache is not shared on ARM CPUs that do not support HT. In the previous implementation of cache passthrough, we assumed that all CPU cores shared the L2 cache. In our tests, we found that the shared settings of the L2 cache will not affect the startup and operation of the VM, but whether it affects the scheduling of the Guest Process is still under further research.

Reply:
Ok, thanks. When I see the 'error!' statement here I thought I cannot launch VM as 'l2_cache_size != 0 && l2_cache_shared' is true in our case.

Reply:
I think the `error!` should be replaced with `warn!`. Because it's not an error really.